classes/steps/actions/webservice_action_step.php

```diff
@@ -0,0 +1,162 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_trigger\steps\actions;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->libdir . '/externallib.php');
+
+/**
+ * Webservice action step class.
+ *
+ * @package    tool_trigger
+ * @author     Kevin Pham <kevinpham@catalyst-au.net>
+ * @copyright  Catalyst IT, 2022
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class webservice_action_step extends base_action_step {
+
+    use \tool_trigger\helper\datafield_manager;
+
+    /** @var string $functionname Name of the function to be called */
+    protected $functionname;
+
+    /** @var int $username The user in which this action will be actioned in the context of */
+    protected $username;
+
+    /** @var int $params parameters that will be used in the corresponding web service function call */
+    protected $params;
+
+    /**
+     * The fields supplied by this step.
+     *
+     * @var array
+     */
+    private static $stepfields = [
+        'data',
+        'error',
+        'exception',
+    ];
+
+    protected function init() {
+        $this->functionname = $this->data['functionname'];
+        $this->username = $this->data['username'];
+        $this->params = $this->data['params'];
+    }
+
+    /**
+     * Returns the step name.
+     *
+     * @return string human readable step name.
+     */
+    public static function get_step_name() {
+        return get_string('webserviceactionstepname', 'tool_trigger');
+    }
+
+    /**
+     * Returns the step name.
+     *
+     * @return string human readable step name.
+     */
+    public static function get_step_desc() {
+        return get_string('webserviceactionstepdesc', 'tool_trigger');
+    }
+
+    /**
+     * Runs the configured step.
+     *
+     * @param $trigger
+     * @param $event
+     * @param $stepresults - result of previousstep to include in processing this step.
+     * @return array if execution was succesful and the response from the execution.
+     */
+    public function execute($step, $trigger, $event, $stepresults) {
+        global $CFG, $DB;
+        require_once($CFG->dirroot . '/admin/tool/trigger/guzzle/autoloader.php');
+
+        $this->update_datafields($event, $stepresults);
+
+        // Passing any data from previous steps through by applying template magic.
+        $username = $this->render_datafields($this->username);
+        $functionname = $this->render_datafields($this->functionname);
+        $params = $this->render_datafields($this->params);
+
+        // Assume the role of the provided user given their {username}.
+        $user = $DB->get_record('user', ['username' => $username], '*', MUST_EXIST);
```

Comment:
How nicely does this fail if the user doesn't exist?

Reply:
wonder about protections around suspended users, but I think that will get caught in the ws internals

Reply:
When a user does not exist, debugging is enabled and the step has "Real time processing" enabled, then it will output the stacktrace to the screen on event trigger. But the rest of the page seems to load okay.

The workflow history stores this error in the step details > step results, below is an example of a non-existent admin:
```
Step results:
'{
    "error": "Invalid user (SELECT * FROM {user} WHERE username = ?\\n[array (\\n  0 => \'admin11\',\\n)])"
}'
```

With the history showing the expected error state for that step.

![image](https://user-images.githubusercontent.com/9924643/157355614-b0c7d1fa-bdb0-44b4-af6c-fc1114285e27.png)